src/modules/did/did.service.ts

```diff
@@ -99,49 +109,124 @@ export class DIDService {
     });
   }
 
+  private parseLogs(logs: string) {
+    const knownBigNumberProperties = new Set<string>();
+    knownBigNumberProperties.add('validity');
+    knownBigNumberProperties.add('topBlock');
+
+    // Operations on IDIDLogData expect some properties to be BigNumbers
+    const bigNumberReviver = (key: string, value) => {
+      if (knownBigNumberProperties.has(key) && '_hex' in value) {
+        return bigNumberify(value['_hex']);
+      }
+      return value;
+    };
+    return JSON.parse(logs, bigNumberReviver) as IDIDLogData;
+  }
+
+  private resolveDocumentFromLogs(id: string, logs: IDIDLogData) {
+    return documentFromLogs(id, [logs]);
+  }
+
+  /**
+   * Reads all logs for a given DID that haven't been cached yet and merged them
+   * @param documentEntity
+   */
+  private async updateLogs(
+    id: string,
+    logs: IDIDLogData,
+  ): Promise<IDIDLogData> {
+    if (logs?.topBlock) {
+      const readFromBlock = logs.topBlock.add(1); // Want to read from 1 more than previously last read to
+      const newLogs = await this.resolver.readFromBlock(id, readFromBlock);
+      const mergedLogs = mergeLogs([logs, newLogs]);
+      mergedLogs.topBlock = newLogs.topBlock;
+      return mergedLogs;
+    }
+    const newLogs = await this.getAllLogs(id);
+    const mergedLogs = mergeLogs([logs, newLogs]);
+    mergedLogs.topBlock = newLogs.topBlock;
+    return mergedLogs;
+  }
+
+  /**
+   * Gets all logs for a given DID
+   * @param documentEntity
+   */
+  private async getAllLogs(id: string): Promise<IDIDLogData> {
+    const genesisBlockNumber = 0;
+    const readFromBlock = utils.bigNumberify(genesisBlockNumber);
+    return this.resolver.readFromBlock(id, readFromBlock);
+  }
+
+  private updateServices(
```

Comment:
Maybe a name like `getMissingClaims` would be more accurate as an `IClaim[]` is being returned and there is no update that is taking place (i.e. no data is actually mutated)

Reply:
👍